pkg/consuldp/lifecycle.go

```diff
@@ -211,3 +216,60 @@ func (m *lifecycleConfig) gracefulShutdown() {
 	// Wait for context timeout to elapse
 	wg.Wait()
 }
+
+func (m *lifecycleConfig) gracefulStartupHandler(rw http.ResponseWriter, _ *http.Request) {
+
+	//Unlike in gracefulShutdown, we want to delay the OK response until envoy is ready
+	//in order to block application container.
+	m.gracefulStartup()
+	rw.WriteHeader(http.StatusOK)
+
+}
+
+// gracefulStartup blocks until the startup grace period has elapsed or we have confirmed that
+// Envoy proxy is ready.
+func (m *lifecycleConfig) gracefulStartup() {
+	timeout := time.Duration(m.startupGracePeriodSeconds) * time.Second
+	m.logger.Info(fmt.Sprintf("blocking container startup until Envoy ready or grace period of %d seconds elapsed", m.startupGracePeriodSeconds))
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	envoyReady := false
+	var wg sync.WaitGroup
+	wg.Add(1)
+	envoyStatus := make(chan bool)
+
+	go func() {
+		defer wg.Done()
+		envoyReady, _ = m.proxy.Ready()
+		envoyStatus <- envoyReady
+
+		//Loop until either proxy is ready or timeout expires.
+	loop:
+		for {
+			select {
+			case <-ctx.Done():
+				break loop
+			case envoyReady = <-envoyStatus:
+				if envoyReady {
+					break loop
+				} else {
+					//Check if Envoy is ready, don't block here so that timeout break can still happen.
+					go func() {
+						ready, _ := m.proxy.Ready()
+						envoyStatus <- ready
+					}()
+				}
+
+			}
+		}
+
+	}()
+
+	wg.Wait()
+
+	if !envoyReady {
+		m.logger.Info("startup grace period reached before envoy ready")
```

Comment:
Is there a `warn` level? Feels like we would want to surface this to the operator.